---
title: "scBubbletree: transparent and quantitative exploration of scRNA-seq data"
author: 
  - Simo Kitanovski, Daniel Hoffmann
  - Bioinformatics and Computational Biophysics, University Duisburg-Essen
output:
  html_document: default
editor_options: 
  markdown: 
    wrap: 72
geometry: margin=1.25in
---




```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = T,
                      comment = F, 
                      warning = F, 
                      message = F)
```

# Before we start we need to install few packages {.unnumbered}

```{r, echo = T, results='hide'}
# load R-package devtools
library(devtools)

# try to install the development version of scBubbletree
devtools::install_github("snaketron/scBubbletree")

# if the installation is a success, load scBubbletree
library(scBubbletree)

# else, ask for help
```


# CITE-seq data 
For this tutorial, we will be analyzing a dataset of Peripheral Blood Mononuclear 
Cells (PBMC) which was obtained from the publication by Hao et al.[^1]. There are 
162,000 single cells in the dataset from eight healthy volunteers enrolled in an 
HIV vaccine trial. This data was collected as part of a larger project in which 
CITE-seq technology[^3] was used with up to 228 antibodies, to simultaneously 
profile single cell gene expressions and cell surface protein expressions. 
Sequencing was done on the Illumina NovaSeq 6000 platform. 

As we have limited time and computational resources for this tutorial, we will 
analyze on a subset of 10,000 randomly selected single cells from three 
volunteers.


## Data processing
Data processing was performed with Seurat. Normalization of gene expressions 
with R-package SCTransform[^2] and dimensionality reduction was performed by 
principal component analysis (PCA) based on the 5,000 most variable genes in 
the dataset. 

We will use the first 15 principal components (first 15 PCs matrix: 
$A^{10,000 \times 15}$) for the downstream analysis, as the first 15 PCs 
captured most of the variance in the data and the proportion of variance 
explained by each subsequent principal component was negligible.

Read and inspect matrix $A^{10,000 \times 15}$ now:

```{r}
# load the data
A <- get(load(file = "case_study_D/A.RData"))
```


```{r}
# inspect A!
#
# what are the dimensions of A? rows -> cells, columns -> PCs
dim(A)
```


## Meta-data
The multimodal data (RNA-seq + surface protein expression) was clustered with 
a weighted nearest neighbor (WNN) approach to identify different cellular 
states in PBMCs based on a weighted combination of both modalities. Cell types 
were predicted by this approach for each cell at three levels of resolution: 
**l1** with 8 cell types; **l2** with 31 cell types; and **l3** with 57 cell 
types. The predictions are available as part of a meta-data associated with the 
dataset.

Read and inspect the meta-data now:

```{r}
meta <- get(load(file = "case_study_D/meta.RData"))
```

```{r}
# inspect meta!
#
# which cell types are available in set l1?
table(meta$celltype.l1)
```


```{r, eval = T}
# ... how about l2? 
table(meta$celltype.l2)
```


```{r, eval = F}
# ... and l3?
table(meta$celltype.l3)

# what other information is included in the meta-data?
# "nCount_RNA" = number of RNA molecules per cell
# "nFeature_RNA" = number of genes/transcripts per cell
# "mt" = percentage of RNAs coming from mitochondrial genes in each cell
# "nCount_ADT" = number of molecules carrying antibodies (ADTs) per cell
# "nFeature_ADT" = number of ADTs (max possible 228 given) per cell
```


## Gene expression data

We have single cell (normalized) expressions for over 20,000 genes in 10,000 
cells. From this data we selected 12 marker genes which should help us identify 
different PBMC subtypes (e.g. B-cells, monocytes, etc.), including:

  * GNLY, NKG7:	*NK cells*
  * IL7R:	*CD4 T cells*
  * CD8A:	*CD8 T cells*
  * MS4A1:	*B cells*
  * CD14, LYZ:	*CD14+ Monocytes*
  * FCGR3A, MS4A7:	*FCGR3A+ Monocytes*
  * FCER1A, CST3:	*Dendritic Cells*
  * PPBP:	*Megakaryocytes*
  
Inspect the normalized gene expression data for these markers now:

```{r}
# load the normalized gene expressions for 12 markers
markers_RNA <- get(load(file = "case_study_D/markers_RNA.RData"))
```


```{r, fig.width=5, fig.height=4, fig.align='center'}
# inspect the data!

# what are the dimensions of the matrix?
dim(markers_RNA)

# which genes are included as columns?
colnames(markers_RNA)

# plot the normalized gene expression distribution for a specific marker 
hist(markers_RNA[, "CD8A"], breaks = 100)
```


## Surface protein expression data
Normalized expression of up to 228 surface antibodies (ADTs) are available 
for each cell:

```{r}
# load the normalized gene expressions for 12 markers
markers_ADT <- get(load(file = "case_study_D/markers_ADT.RData"))
```


```{r, fig.width=5, fig.height=4, fig.align='center'}
# inspect the data!

# what are the dimensions of the matrix?
dim(markers_ADT)

# which genes are included as columns?
colnames(markers_ADT)

# plot the normalized gene expression distribution for a specific marker 
hist(markers_ADT[, "CD8"], breaks = 100)
```

------------------------------------------------------------------------


# Exploration of single cell RNA-seq data with `scBubbletree`

`scBubbletree` is an R-package which provides an easy-to-use workflow for 
transparent and quantitative exploration of single cell RNA-seq data. The
workflow of `scBubbletree` is described by the following figure:

![](graphical_abstract_new.png)


## `scBubbletree` input
The `scBubbletree` workflow takes as first input a matrix $A^{n\times f}$, which 
represents a low-dimensional projection of the high-dimensional scRNA-seq data, 
with $n$ rows as cells and $f$ columns as low-dimension features. 


## `scBubbletree` workflow
In short, `scBubbletree` performs clustering to identify groups of 
transcriptionally similar cells, and then visualizes the identified clusters 
as **bubbles** organized in a hierarchical dendrogram (**bubbletree**) which 
describes the natural (biological) relationships between the bubbles.

The workflow comprises four steps: 

  1. determining the number $k$ of clusters in the data
  2. clustering with k-means
  3. hierarchical cluster grouping
  4. annotation & evaluation
  
  

------------------------------------------------------------------------
  
  
In the following we will analyze $A^{10,000 \times 15}$ based on this workflow.



# (1) determining the number $k$ of clusters in the data 
`scBubbletree` uses k-means for clustering. To apply k-means we must first 
specify the number $k$ of clusters. A choice for $k$ depends on multiple factors. 
First, $k$ is tissue specific as tissues vary in terms of their cell type 
heterogeneity. Second, $k$ is affected by the throughput of the experiment, 
i.e. high throughput experiments (e.g. with 100,000 cells) allow us to extract 
a high-resolution description of the cellular composition of the sample (hence 
high $k$) compared to low-throughput experiments. Third, a choice for $k$ 
depends on the research question. 

In the workflow of `scBubbletree` we propose two general approaches to determine 
$k$: 

First, we should use prior biological knowledge about the cellular composition of 
the tissue (here PBMCs) to determine a potential range of useful values of $k$. 

## PROBLEMS [A]: {.unnumbered}

<hr style="border:3px solid black">

</hr>

1. Can you please investigate the cellular composition of PBMCs based on the 
   human protein atlas (HPA; https://www.proteinatlas.org/humanproteome/immune+cell) 
   or wikipedia (https://en.wikipedia.org/wiki/Peripheral_blood_mononuclear_cell) 
   and suggest a reasonable value for $k$?
   
2. How many major PBMC subtypes are there (**Hint:** approximate number would 
   suffice here)?

------------------------------------------------------------------------

Second, `scBubbletree` offers the function `get_k` to estimate $k$ from the 
data using the *Gap statistic* and the within-cluster variance (within cluster 
sum of squares; *WCSS*). These data-driven metrics describe quantitatively the 
extent to which different observations fall into natural distinct clusters as 
a function of $k$

------------------------------------------------------------------------

## PROBLEMS [B]: {.unnumbered}

<hr style="border:3px solid black">

</hr>

1. Please run `get_k` now: 

```{r, fig.width=4, fig.height=3, echo = T, eval=F}
# Executing this command can take couple of minutes. You can instead load 
# the precomputed results (see the next line of code)

b <- scBubbletree::get_k(x = A, 
                         B = 3,
                         cv_prop = 0.5,
                         ks = seq(from = 1, to = 45, by = 3),
                         kmeans_algorithm = "MacQueen",
                         cores = 4)

# Parameter description:
# x: input matrix A
# B: number of bootstrap iterations
# cv_prop: proportion of the data to be use in each bootstrap iteration
# ks: vector of k values to consider
# kmeans_algorithm: one of few k-means algorithm variants
# cores: umber of PC cores to use in parallel
```

```{r, echo=F, eval=F}

b <- scBubbletree::get_k(B = 3,
                         cv_prop = 0.5,
                         ks = seq(from = 1, to = 45, by = 3),
                         x = A,
                         cores = 4,
                         kmeans_algorithm = "MacQueen")

save(b, file = "case_study_D/b.RData")

```

```{r, eval=T, echo=F}
# load precomputed 'b'
b <- get(load(file = "case_study_D/b.RData"))
```


2. Next, plot the Gap statistic and the WCSS curves. Do you see an `elbow` in 
   either curve? 
   
   (**Hint**: elbow or knee is a point where the curve visibly bends, 
   specifically from high slope to low slope. We are interested in finding $k$
   that is as near as possible to the elbow/knee point.)
   
```{r, fig.width=8, fig.height=4, fig.align='center'}
ggplot(data = b$gap_stats_summary)+
  geom_point(aes(x = k, y = gap_mean), size = 1)+
  geom_errorbar(aes(x = k, y = gap_mean, ymin = L95, ymax = H95), width = 0.1)+
  ylab(label = "Gap")|
ggplot(data = b$wss_stats_summary)+
  geom_point(aes(x = k, y = wss_mean), size = 1)+
  geom_errorbar(aes(x = k, y = wss_mean, ymin = L95, ymax = H95), width = 0.1)+
  ylab(label = "WCSS")+
  scale_y_log10()+
  annotation_logticks(base = 10, sides = "l")
```


3. Is the $k$ value that is found near the `elbow` approximately equal to the 
   number of major cell subtypes in PBMCs?

------------------------------------------------------------------------

# (2) clustering with k-means and (3) hierarchical cluster grouping
We have established that $18$ is a reasonable first choice for $k$. Lets perform
k-means clustering with $k=18$ next:

```{r, echo=T, eval = F}

# Executing this command can take couple of minutes. You can instead load the
# precomputed resulta (see the next line of code)

k18 <- scBubbletree::get_bubbletree(x = A,
                                    k = 18,
                                    seed = 241142,
                                    cores = 4,
                                    N_eff = 100,
                                    show_simple_count = F,
                                    round_digits = 1,
                                    kmeans_algorithm = "MacQueen")


# Parameter description:
# A: input matrix
# k: selected k
# seed: use seed for reproducible results
# cores: number of PC cores to use in parallel
# B: number of bootstrap iterations -> dendrograms
# N_eff: number of bubble cells to use when computing inter-cluster distances
# show_simple_count: divide cell counts in bubbles by 1000 (for aesthetics)
# round_digits: round cell counts (for aesthetics)
```

```{r, echo=F, eval=F}
save(k18, file = "case_study_D/k18.RData")
```

```{r, echo=F, eval=T}
k18 <- get(load(file = "case_study_D/k18.RData"))
```


## PROBLEMS [C]: {.unnumbered}

<hr style="border:3px solid black">

</hr>

1. Visualize the bubbletree

```{r, eval=T, echo=T, fig.width=6, fig.height=5}
k18$tree
```

2. Which are the largest/smallest bubbles in the bubbletree? How many cells
   do they contain?
   
   
3. About 25% of all cells belong to bubble 14. Can you guess the PBMC subtype 
   of the cells occupying this bubble?

  (*Hint:* Which major PBMC subtype constitutes the largest proportion of PBMCs? 
  https://en.wikipedia.org/wiki/Peripheral_blood_mononuclear_cell)


4. How many major clades do you see in the bubbletree? Based on the prior 
   knowledge on PBMCs can you guess which major PBMC subtypes are contained 
   in these clades?


5. Do all branches have high degree of support?

  (**Hint:** what is the meaning of the red labels shown on top of the different 
  branches)?

------------------------------------------------------------------------

# (4) annotation and model evaluation
We need to 'attach' annotations to the bubbletree to make it biologically 
explainable. 



## "Attaching" marker gene expressions

Gene expressions are **numeric** cell features. `scBubbletree` offers two 
functions to visualize numeric cell features.

 * Function `get_num_feature_tiles`: visualizes summaries (e.g. average gene 
   expressions) in each bubble using tile plots, where each tile is color-coded 
   and labeled with the average value of a specific feature in a given bubble.
 
 * Function `get_num_feature_violins`: visualizes the feature distribution in 
   each bubble using violin plots.

Lets use `get_num_feature_tiles` to visualize the average (normalized) gene 
expressions of 12 markers:

```{r}
markers_RNA_plot <- scBubbletree::get_num_feature_tiles(d = k18,
                                                        as = markers_RNA,
                                                        round_digits = 1)
```


```{r, fig.width=8, fig.height=6, fig.align='center', echo=T}
# Markers and their cell types:
# * GNLY, NKG7:	NK cells
# * IL7R:	CD4 T cells
# * CD8A:	CD8 T cells
# * MS4A1:	B cells
# * CD14, LYZ:	CD14+ Monocytes
# * FCGR3A, MS4A7:	FCGR3A+ Monocytes
# * FCER1A, CST3:	Dendritic Cells
# * PPBP:	Megakaryocytes

(k18$tree | markers_RNA_plot$w)+
  patchwork::plot_layout(widths = c(1, 1))+
  patchwork::plot_annotation(tag_levels = 'A')
```


## PROBLEMS [D]: {.unnumbered}

<hr style="border:3px solid black">

</hr>

1. What is the most likely cell type of the cells occupying:
    a. bubbles 1 and 17
    b. bubbles 4, 10, 12
    c. bubbles of the top major clade (1, 17, 10, 4, 12, 7, 14, 11, 18)
    d. bubbles of the bottom major clade (8, 5, 2, 13, 9 and 15)
    e. bubble 16
    

2. Does the topology of the bubbletree make sense from a biological 
   point of view?


------------------------------------------------------------------------


## "Attaching" surface-protein expressions

We can also use the surface protein expression to better understand the 
bubbletree. Lets visualize the average surface protein expressions for five 
markers using `get_num_feature_tiles`:

  * CD3-1 = T cells
  * CD19 = B cells
  * CD14 = CD14+ monocytes
  * CD56-1 = NK cells
  * CD11c = dendritic cells (DCs) and monocytes


```{r}

markers_ADT_plot <- scBubbletree::get_num_feature_tiles(d = k18,
                                                        as = markers_ADT[, c("CD3-1",
                                                                             "CD19", 
                                                                             "CD14", 
                                                                             "CD56-1",
                                                                             "CD11c")],
                                                        round_digits = 1)
```



```{r, fig.width=9, fig.height=6, fig.align='center', echo=T}
# combine 3 plots with patchwork: bubbletree + rna-markers + surface-protein-markers

(k18$tree | markers_RNA_plot$w | markers_ADT_plot$w)+
  patchwork::plot_layout(widths = c(1.5, 2, 1))+
  patchwork::plot_annotation(tag_levels = 'A')
```


## PROBLEMS [E]: {.unnumbered}

<hr style="border:3px solid black">

</hr>

1. Are the results obtained based on the different modalities (RNA-seq vs. 
   surface protein expression) in agreement with each other?

2. Please add CD8 to the list of surface proteins and visualize its mean 
   expression with `get_num_feature_tiles` to find out which bubbles contain 
   CD8+ T cells    




## "Attaching" cell type predictions (**l1** and **l2** set)

Cell can have **categorical** features. For instance:

  * predicted cell types (CD4, CD8, B-cell, ...)
  * cell cycle phase (G1, S, G2, ...)
  * sample name (P1, P2, P3, ...)
  * biological condition (e.g. infected vs. control)
  * singlet vs. multiplet
  * ...
 
`scBubbletree` can visualize categorical features using tile plots with 
the function `get_cat_feature_tiles`. With the help of the Boolean input 
parameter `integrate_vertical` the tiles in the tile plot can be configured 
for two types of interpretation: 

  * With `integrate_vertical=TRUE` the tiles can be configured to show the 
  relative frequencies of cells carrying a specific label/category of the 
  categorical feature across the different bubbles. Columns (features) in the 
  tile plot will integrate to 100%. With this type of visualization we can answer 
  questions such as: are cells with a specific label enriched in a particular bubble? 

  * With `integrate_vertical=FALSE` we compute the relative frequencies of the 
  different labels within in each bubble, i.e. the feature composition of the 
  different bubbles. Rows (bubbles) in the tile plot will integrate to 100%. With 
  this we can answer questions such as: how 'pure' is a given bubble, i.e.\ is 
  the bubble composed of cells that only carry a particular label or a mixture 
  of different labels? 

Lets visualize the **l1** set of cell type predictions. We will set 
`integrate_vertical=TRUE` and `integrate_vertical=FALSE` and show both results:

```{r}
celltype_l1_vertical <- scBubbletree::get_cat_feature_tiles(d = k18,
                                                    a = meta$celltype.l1,
                                                    integrate_vertical = T,
                                                    round_digits = 1,
                                                    rotate_x_axis = T)

celltype_l1_horizontal <- scBubbletree::get_cat_feature_tiles(d = k18,
                                                    a = meta$celltype.l1,
                                                    integrate_vertical = F,
                                                    round_digits = 1,
                                                    rotate_x_axis = T)
```

```{r, fig.width=9, fig.height=6, fig.align='center', echo=T}
(k18$tree | celltype_l1_vertical$w | celltype_l1_horizontal$w)+
  patchwork::plot_layout(widths = c(1, 1, 1))+
  patchwork::plot_annotation(tag_levels = 'A')
```

Some observations about the above figure:

  * we see enrichment of cells from a specific cell type in either one bubble 
    or few bubbles that are typically found in proximity of each other in the 
    bubbletree (Fig. B). For instance, about 98\% of NK cells in the sample are 
    enriched in bubbles 9, 4 and 14 (Fig. B), while about 98\% of B cells are 
    enriched in bubbles 11 and 8 (Fig. B). Some cell types, such as monocytes 
    and T cells, are spread across more than two bubbles.
    
  * we confirm that the top and bottom major clades of the bubbletree are formed 
  by bubbles that are enriched with lymphocytes and monocytes, respectively. The 
  remaining outgroup bubbles contain a mixture of cells termed as 'other', DCs 
  and a small multiple other cells. 
  
  * within the lymphocyte clade the branching adequately captures the relationship 
  between the major lymphocyte subtypes, placing T cells, B cells and NK cells 
  into their own robust subclades. 
  
  * the fact that monocytes occupy multiple bubbles of the bottom major clade 
  suggests that there might exists some degree of heterogeneity within the 
  monocyte class that is captured by the different bubbles. It is also possible 
  that the selected value for $k$ is too high leading to oversegmentation of 
  large but rather homogenous populations. To check which of these claims is 
  more likely, we need to ``attach'' to the bubbletree cell type predictions 
  at a higher resolution. We show this in the following.
  
  * using Fig. B and C in combination with the bubbletree allows us to get a 
  clear sense of how the cells are distributed across the bubbles and about the 
  'purity' of the bubbles. From this we can detect abnormalities at an early
  stage of the scRNA-seq data analysis



## PROBLEMS [F]: {.unnumbered}

<hr style="border:3px solid black">

</hr>

1. Do the same analysis for the set **l2**

2. We know that the bubbles 8 and 11 contain B cell bubbles. Is this 
   partitioning of the B cell population biologically sound?



```{r, echo=F}
celltype_l2_vertical <- scBubbletree::get_cat_feature_tiles(d = k18,
                                                    a = meta$celltype.l2,
                                                    integrate_vertical = T,
                                                    round_digits = 0,
                                                    rotate_x_axis = T)

celltype_l2_horizontal <- scBubbletree::get_cat_feature_tiles(d = k18,
                                                    a = meta$celltype.l2,
                                                    integrate_vertical = F,
                                                    round_digits = 0,
                                                    rotate_x_axis = T)
```

```{r, fig.width=15, fig.height=6, fig.align='center', echo=F}
(k18$tree | celltype_l2_vertical$w | celltype_l2_horizontal$w)+
  patchwork::plot_layout(widths = c(1, 3, 3))+
  patchwork::plot_annotation(tag_levels = 'A')
```



## Data quality control with `scBubbletree` 

### Compositional checks
In the analysis of scRNA-seq datasets from multiple donors or biological 
conditions we want to check if there are systematic differences in cell type 
composition between the different samples. Finding such compositional 
differences may help us identify novel cell types (e.g.tumor cells) that occur 
in some of the samples but not in others, or to detect problematic samples with 
biased cell type compositions caused by e.g. flaw in the experiment. 

As donor/sample IDs are categorical cell features, we can used the function 
`get_cat_feature_tiles` to visualize the relative frequency of cells from a 
given donor across the different bubbles

Lets do this now:

```{r, fig.width=7, fig.height=6, fig.align='center', eval=T, echo = T}
donor_features <- scBubbletree::get_cat_feature_tiles(d = k18,
                                                      a = meta$donor,
                                                      integrate_vertical = T,
                                                      round_digits = 1)
```


```{r, fig.width=7, fig.height=6, fig.align='center', eval = T}
(k18$tree | donor_features$w)+
  patchwork::plot_layout(widths = c(1, 1))
```

## PROBLEMS [G]: {.unnumbered}

<hr style="border:3px solid black">

</hr>

1. Do you see similar relative frequencies (here percentages) of P1, P2 and P3 
   cells across the different bubbles?

**Important remark**: with this `scBubbletree` provides a quick qualitative 
description of the relative frequencies of each cell feature [here volunteer 
name] which is essential to have at the early stage of a scRNA-seq data analysis. 
Furthermore, `scBubbletree` provides the raw data in the form of a table with 
cell counts for each pair of feature and bubble, which can be used as input for 
quantitative differential abundance analysis of cell type composition[^4].

```{r}
# the object donor_features contains the raw data

knitr::kable(x = head(donor_features$ws[, c("cluster", "feature", "freq")]),
             row.names = F)
```


### Other checks
Other quality checks are also possible with `scBubbletree`. This includes 
checks for bubbles that are enriched with cells with high expression levels 
of mitochondrial genes, which is an indicator of poor sample quality associated 
with high fraction of apoptotic or lysing cells; or cells having low overall 
number of genes or RNA molecules detected.

Lets plot the distributions of these *numeric feature* in each bubble:

```{r, fig.width=9, fig.height=5, fig.align='center', echo=T}
qc <- scBubbletree::get_num_feature_violins(d = k18,
                                            as = meta[, c("nCount_RNA",
                                                          "nFeature_RNA",
                                                          "mt")],
                                            show_cells = F)
```


```{r, fig.width=8, fig.height=6, fig.align='center', echo=T}
(k18$tree | qc$w)+
  patchwork::plot_layout(widths = c(1.25, 2))
```


## PROBLEMS [H]: {.unnumbered}

<hr style="border:3px solid black">

</hr>

1. Do you notice any serious problems with the data? 

(**Hint:** notice the conspicuously low number of genes or RNA molecules 
in bubble 15. The data should be devoid of major quality problems as it has 
already been processed by Hao et al.[^1])










[^1]: Hao, Yuhan, et al. "Integrated analysis of multimodal single-cell data." 
Cell 184.13 (2021): 3573-3587.
[^2]: Hafemeister, Christoph, and Rahul Satija. "Normalization and variance 
stabilization of single-cell RNA-seq data using regularized negative binomial 
regression." Genome biology 20.1 (2019): 1-15.
[^3]: Stoeckius, Marlon, et al. "Simultaneous epitope and transcriptome 
measurement in single cells." Nature methods 14.9 (2017): 865-868.
[^4]: Buettner, Maren, et al. "scCODA is a Bayesian model for compositional 
single-cell data analysis." Nature communications 12.1 (2021): 1-10.
